types/coin.go

```diff
@@ -270,6 +282,10 @@ func (coins Coins) Add(coinsB ...Coin) Coins {
 // denomination and addition only occurs when the denominations match, otherwise
 // the coin is simply added to the sum assuming it's not zero.
 func (coins Coins) safeAdd(coinsB Coins) Coins {
+	if !coinsB.isSorted() {
```

Comment:
What is the reasoning for not checking if `coins` are Sorted?

Reply:
why not just sort the coins instead of a panic? You are returning a new coin set anyways since the coins are being added or subtracted

Reply:
Thanks for review. We can create a new array and sort it. I wanted to avoid sorting in place to not change the argument - when calling `coinsA.Add(coinsB)`, it could be surprising for a user that we are changing the argument (`coinsB`).

Reply:
The main reason I'm panicing is because other methods are doing the same in `coins`.

Reply:
> What is the reasoning for not checking if coins are Sorted?

See the NOTE in the PR description. I can add it as well.
For panic vs error: checking if list is sorted is faster than copying it and sorting.

Reply:
The reasoning for panicing makes sense. Lets stick with that. To add to the argument, if coins are being added/subtracted that are unsorted, these are considered invalid coin sets and throwing a panic makes the most sense (as the code should be fixed immediately)

Reply:
> The reasoning for panicing makes sense. Lets stick with that. To add to the argument, if coins are being added/subtracted that are unsorted, these are considered invalid coin sets and throwing a panic makes the most sense (as the code should be fixed immediately)

I agree. The code must `panic()` when something entirely unexpected and unrecoverable happens.

Reply:
> What is the reasoning for not checking if coins are Sorted?

> it seams we construct it correctly with a constructor (NewCoins).

Cosmos SDK users that are unaware that coins are sorted, might still do `sdk.Coins{coinC, coinB, coiniA}`. So I think checking sort on self is also useful.

Reply:
Probably the best way would be to make the type private, and only expose an interface. I will add a check to self and note to hide the type in the future.

Reply:
> (Newer) cosmos SDK users that are unaware that coins are sorted, might still do `sdk.Coins{coinC, coinB, coiniA}`

Yes this case can be exist.

> So I think checking sort on self is also useful.

I agree.